tfhe/src/integer/bigint/u512.rs

```diff
@@ -0,0 +1,362 @@
+        super::algorithms::bitand_assign(self.0.as_mut_slice(), rhs.0.as_slice())
+    }
+}
+
+impl std::ops::BitOrAssign<Self> for U512 {
+    fn bitor_assign(&mut self, rhs: Self) {
+        super::algorithms::bitor_assign(self.0.as_mut_slice(), rhs.0.as_slice())
+    }
+}
+
+impl std::ops::BitOr<Self> for U512 {
+    type Output = Self;
+
+    fn bitor(mut self, rhs: Self) -> Self::Output {
+        self |= rhs;
+        self
+    }
+}
+
+impl std::ops::BitXorAssign<Self> for U512 {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        super::algorithms::bitxor_assign(self.0.as_mut_slice(), rhs.0.as_slice())
+    }
+}
+
+impl std::ops::BitXor<Self> for U512 {
+    type Output = Self;
+
+    fn bitxor(mut self, rhs: Self) -> Self::Output {
+        self ^= rhs;
+        self
+    }
+}
+
+impl From<u8> for U512 {
+    fn from(value: u8) -> Self {
+        Self::from(value as u128)
+    }
+}
+
+impl From<u16> for U512 {
+    fn from(value: u16) -> Self {
+        Self::from(value as u128)
+    }
+}
+
+impl From<u32> for U512 {
+    fn from(value: u32) -> Self {
+        Self::from(value as u128)
+    }
+}
+
+impl From<u64> for U512 {
+    fn from(value: u64) -> Self {
+        Self::from(value as u128)
+    }
+}
+
+impl From<u128> for U512 {
+    fn from(value: u128) -> Self {
+        Self([
+            (value & u128::from(u64::MAX)) as u64,
+            (value >> 64) as u64,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+        ])
+    }
+}
+
+impl CastFrom<U512> for u64 {
+    fn cast_from(input: U512) -> Self {
+        input.0[0]
+    }
+}
+
+impl CastFrom<U512> for u8 {
+    fn cast_from(input: U512) -> Self {
+        input.0[0] as u8
+    }
+}
+
+impl CastFrom<crate::integer::U256> for U512 {
+    fn cast_from(input: crate::integer::U256) -> Self {
+        Self([input.0[0], input.0[1], input.0[2], input.0[3], 0, 0, 0, 0])
+    }
+}
+
+impl CastFrom<u32> for U512 {
+    fn cast_from(input: u32) -> Self {
+        Self::from(input)
+    }
+}
+
+impl CastFrom<u64> for U512 {
+    fn cast_from(input: u64) -> Self {
+        Self::from(input)
+    }
+}
+
+impl CastFrom<u8> for U512 {
+    fn cast_from(input: u8) -> Self {
+        Self::from(input as u64)
+    }
+}
+
+impl From<bool> for U512 {
+    fn from(input: bool) -> Self {
+        Self::from(if input { 1u64 } else { 0u64 })
+    }
+}
+
+// SAFETY
+//
+// U512 is allowed to be all zeros
+unsafe impl bytemuck::Zeroable for U512 {}
+
+// SAFETY
+//
+// u64 impl bytemuck::Pod,
+// [T; N] impl bytemuck::Pod if T: bytemuck::Pod
+//
+// https://docs.rs/bytemuck/latest/bytemuck/trait.Pod.html#foreign-impls
+//
+// Thus U512 can safely be considered Pod
+unsafe impl bytemuck::Pod for U512 {}
+
+impl Numeric for U512 {
+    const BITS: usize = Self::BITS as usize;
+
+    const ZERO: Self = Self::ZERO;
+
+    const ONE: Self = Self::ONE;
+
+    const TWO: Self = Self::TWO;
+
+    const MAX: Self = Self::MAX;
+}
```

Comment:
no tests for u512 inspired of u256 module ?

Reply:
I added some tests for u512